### ISSUE TITLE
Training on Amazon EC2 on Windows

### ISSUE THREAD

**Issue by silvernine209** [AlexeyAB/darknet#1380]
I have some credit in AWS EC2 and would like to train using it.
I did some research, but I'm still clueless how I can utilize AWS EC2 and train my darknet on my Windows setup...any help please?

**Comment by PeterQuinn925** [comment]
I haven't used AWS in a while but last time I did, it was just like any other VM. Just install darknet and train. What kind of help are you looking for?

**Comment by silvernine209** [comment]
@PeterQuinn925 Thank you for the reply. I have 1.6 milion images with 500 categories to train for. I've already trained up to 43,000 iterations on my local machine, and looking for a lot more iterations.

I'm looking to continue training using AWS instance, does this mean I have to upload or move all of my weights, images, repo, and etc to the VM you are talking about?

I was anticipating this to happen with AWS instance:
    1.) Initiate AWS E3 instance
    2.) Connect to the E3 instance on my command prompt on my Windows
    3.) Begin training using GPU from E3 instance using "darknet.exe detector train data/obj.data yolo-obj.cfg backup-transfer_learning/yolo-obj_43000.weights"

I feel like I'm missing something fundamental..

**Comment by PeterQuinn925** [comment]
Yeah. I think you do need to copy the exe, cfg, weights, images to your AWS instance.  Perhaps someone with experience with doing this on AWS can answer.

**Comment by AlexeyAB** [comment]
To train on Amazon AWS EC2:

* Use free instance `t2.micro` and `AMI (Ubuntu) Version 3.0 (ami-38c87440)` - Deep Learning base AMI with NVidia drivers like CUDA 8 and 9, CuDNN 6 and 7
* Attach this AMI to the t2.micro
* Upload  from your computer to the amazon cloud:
    * your dataset (1.6 images) 
    * pre-trained weights (yolo_43000.weights)
    * Darknet
* De-attach this AMI/Volume (with dataset, weights, Darknet)
* Select new instance with GPU - for example  `p3.2xlarge` (Tesla V100), but you may need to contact support to set a limit
* Attach this AMI/Volume (with dataset, weights, Darknet) to this new instance p3.2xlarge
* Compile Darknet with `GPU=1 CUDNN=1 CUDNN_HALF=1 OPENCV=1` in the Makefile


----

For connection to the Amazon EC2 runing instance - you can use Putty: https://putty.org.ru/download.html

* Copy file `yolo_43000.weights` from your computer to the EC2:
`pscp -i t1-free.ppk yolo_43000.weights ubuntu@ec2-35-160-228-91.us-west-2.compute.amazonaws.com:/home/ubuntu/`

* Copy file `dataset.zip` from your computer to the EC2:
`pscp -i t1-free.ppk dataset.zip ubuntu@ec2-35-160-228-91.us-west-2.compute.amazonaws.com:/home/ubuntu/`

* Use `putty.exe` and your Amazon EC2 private key (for example `t1-free.ppk`) to connect to the `t2.micro` Linux console

**Comment by silvernine209** [comment]
@AlexeyAB i just open my jaw in amazemament of such wonderful help. Thank you!!

**Comment by silvernine209** [comment]
@AlexeyAB i just open my jaw in amazemament of such wonderful help. Thank you!! I will give it a try today and report on how it goes.

**Comment by AlexeyAB** [comment]
@silvernine209 

* for `p2.xlarge` (Tesla K80) use `GPU=1 CUDNN=1 CUDNN_HALF=0 OPENCV=0` in the Makefile and train with `-gpus 0,1` flag after 1000 iterations

* for `p3.2xlarge` (Tesla V100) use `GPU=1 CUDNN=1 CUDNN_HALF=1 OPENCV=1` in the Makefile and train with `-dont_show` flag

**Comment by silvernine209** [comment]
@AlexeyAB Alexey thank you for your follow up comments. On top of your thorough explanations, I also went through Amazon's tutorials ([link](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/EC2_GetStarted.html?icmpid=docs_ec2_console)). I'm currently uploading my train train set to my virtual machine. After this, I will upload darknet and my weight. 
I shouldn't have any problem from this point, and thank!

**Comment by silvernine209** [comment]
@AlexeyAB p2.xlarge is $0.90/hour and p3.2xlarge is $3.06 per hour. Would you train 3.06/0.9 = 3.4 times faster with p3.2xlarge? If not, I will just stick with p2.xlarge.

Thanks

**Comment by AlexeyAB** [comment]
@silvernine209 

* **p2.xlarge** is $0.90/hour - GPU nVidia K80 - **8 736 Gflops**: https://en.wikipedia.org/wiki/List_of_Nvidia_graphics_processing_units#Tesla

* **p3.2xlarge** is $3.06/hour - GPU nVidia V100 - **134 092 Gflops** if is used `GPU=1 OPENCV=1 CUDNN=1 CUDNN_HALF=1` (like Titan V with Tensor Cores): https://en.wikipedia.org/wiki/List_of_Nvidia_graphics_processing_units#Volta_series

So theoretically p3.2xlarge  is **15x** times faster than p2.xlarge. I didn't test K80, but I think it is about 5-8x times. A little bit about Titan V100 p3.2xlarge: https://github.com/AlexeyAB/darknet/issues/407

Also follow my recommendations for the 1st 1000 iterations: https://github.com/AlexeyAB/darknet/issues/1380#issuecomment-412380005

**Comment by silvernine209** [comment]
@AlexeyAB Should I use `-gpus 0,1` flag after 1000 iterations even for `p3.2xlarge`?

**Comment by AlexeyAB** [comment]
@silvernine209 

* **p2.xlarge** (2xGPU) -  train 1st 1000 iterations without `-gpus 0,1` then continue with `-gpus 0,1` (don't use `CUDNN_HALF=1` for p2.xlarge)

* **p3.2xlarge** (1xGPU) - train 1st 1000 iterations with `CUDNN_HALF=0`, then continue with `CUDNN_HALF=1`

* **p3.8xlarge** (4xGPU) - train 1st 1000 iterations with `CUDNN_HALF=0` and without `-gpus 0,1,2,3`, then continue with `CUDNN_HALF=1` and with flag `-gpus 0,1,2,3`

**Comment by silvernine209** [comment]
@AlexeyAB Thank you!!

**Comment by silvernine209** [comment]
@AlexeyAB I've been trying to train **p2.xlarge** with `-gpus 0,1` and I kept getting this error

![image](https://user-images.githubusercontent.com/37491067/44240327-8fce7b00-a18b-11e8-85ec-88d61c478e97.png)
I think **p2.xlarge** is just one gpu not two.

![image](https://user-images.githubusercontent.com/37491067/44240347-a674d200-a18b-11e8-8cd6-38747e29af60.png)

![image](https://user-images.githubusercontent.com/37491067/44240354-b2609400-a18b-11e8-8415-31339ac5cc33.png)
[link](https://aws.amazon.com/blogs/aws/new-p2-instance-type-for-amazon-ec2-up-to-16-gpus/)

**Comment by AlexeyAB** [comment]
@silvernine209 Yes, It looks like that `p2.xlarge` uses only half of K80 GPU that contains 2 chips GK210.
I fixes my post.

**Comment by mmartin56** [comment]
Hi Alexey, thanks for your valuable help here. I'm using p3.2xlarge (Tesla V100). I have compiled darknet with Makefile settings GPU=1 CUDNN=1 CUDNN_HALF=1 OPENCV=1 and everything runs smoothly. However when instead of modifying the Makefile I set these values in command line (i.e. _make GPU=1 CUDNN=1 CUDNN_HALF=1 OPENCV=1_), I get errors both during training:
![Selection_064](https://user-images.githubusercontent.com/55649430/65649008-27b5be00-e048-11e9-844a-5f8655ad86d1.jpg)
and testing. Testing with debug gives:
![Selection_065](https://user-images.githubusercontent.com/55649430/65649136-9d218e80-e048-11e9-838b-c7aac49957da.jpg)

I would prefer to not modify the Makefile, if possible. Do you know why I get an error when passing options as command line arguments?

**Comment by matt-sharp** [comment]
> @silvernine209
> 
> * **p2.xlarge** (2xGPU) -  train with `CUDNN_HALF=0`
> * **p3.2xlarge** (1xGPU) - train 1st 1000 iterations with `CUDNN_HALF=0`, then continue with `CUDNN_HALF=1`
> * **p3.8xlarge** (4xGPU) - train 1st 1000 iterations with `CUDNN_HALF=0` and without `-gpus 0,1,2,3`, then continue with `CUDNN_HALF=1` and with flag `-gpus 0,1,2,3`

@AlexeyAB If we are expected to change CUDNN_HALF after 1000 iterations how is the best way to go about that in practice?
Should we set max iterations to 1000 in the cfg file initially and then change it?
Do we need to re-compile darknet after 1000 iterations before resuming training?
Is there a better approach rather than making these changes manually?

**Comment by avn3r** [comment]
> > Also, Although I set the learning_rate as 0.001, it is changing to 0.00003 or some different values around 100 iterations.
> 
> Its burn_in (warm up) for the first 1000 iterations, don't wory.
> 
> for training on 4x GPUs:
> 
> * sometimes better to set values `batch/4`, `subdivisions/4`, `learning_rate/4`, `max_batches*4`, `steps*4`
> * sometimes just train as usual

Thanks, @AlexeyAB!

Shouldn't batch size be bigger given that we now using 4 GPUs usually for data-parallel multi-GPU training I would expect `batch=64*4` instead of `batch=64/4`. Am I missing something? Thanks for the clarification.
